jax/experimental/jax2tf/impl_no_xla.py

```diff
@@ -217,19 +232,27 @@ def _conv_general_dilated(
     preferred_element_type: Optional[DType],
     _in_avals: Sequence[core.ShapedArray], _out_aval: core.ShapedArray):
   """Implementation of lax.conv_general_dilated_p using XlaConv."""
+  # In presence of shape polymorphism, lhs.shape and rhs.shape may contain
+  # None. The actual dimension polynomial shapes are in _in_avals.
   del lhs_shape, rhs_shape, precision  # Unused arguments.
-  out_shape = jax2tf._aval_to_tf_shape(_out_aval)
-  _validate_spatial_dimensions(len(lhs.shape) - 2)
-  is_conv1d = len(lhs.shape) - 2 == 1
+  lhs_shape, rhs_shape = _in_avals[0].shape, _in_avals[1].shape
+  jax2tf._assert_matching_abstract_shape(lhs, lhs_shape)
+  jax2tf._assert_matching_abstract_shape(rhs, rhs_shape)
+  out_shape = _out_aval.shape
+  _validate_spatial_dimensions(len(lhs_shape) - 2)
+  is_conv1d = len(lhs_shape) - 2 == 1
 
   tf_window_strides = _normalize_window_strides(window_strides)
   padding, lhs_dilation, rhs_dilation = _normalize_padding_and_dilations(
       padding, lhs_dilation, rhs_dilation, is_conv1d)
 
-  lhs, rhs = _transpose_for_tf_conv(lhs, rhs, dimension_numbers)
-
-  in_channels = lhs.shape[-1]
-  *rhs_spatial_shapes, _, rhs_out_channel = rhs.shape
+  lhs, lhs_shape, rhs, rhs_shape = _transpose_for_tf_conv(lhs, lhs_shape,
+                                                          rhs, rhs_shape,
+                                                          dimension_numbers)
+  jax2tf._assert_matching_abstract_shape(lhs, lhs_shape)
```

Comment:
Looks like you could move these asserts into the _transpose_for_tf_conv function to save some logic in the main branch.